phase4-euctp-client/src/main/java/com/helger/phase4/euctp/AbstractEuctpUserMessageBuilder.java

```diff
@@ -0,0 +1,126 @@
+package com.helger.phase4.euctp;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import javax.annotation.OverridingMethodsMustInvokeSuper;
+
+import com.helger.phase4.crypto.ECryptoAlgorithmC14N;
+import com.helger.phase4.crypto.ECryptoKeyEncryptionAlgorithm;
+import com.helger.phase4.crypto.ECryptoKeyIdentifierType;
+import com.helger.phase4.profile.euctp.EuCtpPMode;
+import com.helger.phase4.sender.AbstractAS4UserMessageBuilderMIMEPayload;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract euctp UserMessage builder class with sanity methods
+ *
+ * @param <IMPLTYPE> The implementation type
+ * @author Ulrik Stehling
+ */
+public abstract class AbstractEuctpUserMessageBuilder<IMPLTYPE extends AbstractEuctpUserMessageBuilder<IMPLTYPE>>
+		extends
+		AbstractAS4UserMessageBuilderMIMEPayload<IMPLTYPE>
+{
+	private static final Logger LOGGER = LoggerFactory.getLogger (AbstractEuctpUserMessageBuilder.class);
+
+	// Default per section 2.2.6.2.1
+	public static final ECryptoKeyIdentifierType DEFAULT_KEY_IDENTIFIER_TYPE_SIGN = ECryptoKeyIdentifierType.BST_DIRECT_REFERENCE;
+	// Default per section 2.2.6.2.2
+	public static final ECryptoKeyIdentifierType DEFAULT_KEY_IDENTIFIER_TYPE_CRYPT = ECryptoKeyIdentifierType.SKI_KEY_IDENTIFIER;
+
+	protected AbstractEuctpUserMessageBuilder()
+	{
+		// Override default values
+		try
+		{
+//        httpClientFactory(new Phase4EuCtpHttpClientSettings());
+
+			// Other crypt parameters are located in the PMode security part
+			cryptParams().setKeyIdentifierType(DEFAULT_KEY_IDENTIFIER_TYPE_CRYPT);
+			cryptParams().setKeyEncAlgorithm(ECryptoKeyEncryptionAlgorithm.ECDH_ES_KEYWRAP_AES_128);
+			cryptParams().setEncryptSymmetricSessionKey(false);
+
+			/*
+			 * Assumption: the BST "ValueType" attribute is set to
+			 * "http://docs.oasis-open.org/wss/2004/01/oasis-200401-wss-x509-token-profile-1.0#X509PKIPathv1"
+			 * by WSS4J automatically (see WSSecSignature#addBST)
+			 */
+
+			// Other signing parameters are located in the PMode security part
+			signingParams().setKeyIdentifierType(DEFAULT_KEY_IDENTIFIER_TYPE_SIGN);
+			signingParams().setAlgorithmC14N(ECryptoAlgorithmC14N.C14N_EXCL_OMIT_COMMENTS);
+			// Use the BST value type "#X509PKIPathv1"
+			signingParams().setUseSingleCertificate(false);
+
+			conversationID("1");
```

Comment:
A quick explanation would be nice why this must be `1`

Reply:
Changed to random UUID